odoo/models.py

```diff
@@ -3476,37 +3478,110 @@ def _check_parent_path(self):
             _logger.error('parent_path field on model %r should be indexed! Add index=True to the field definition.', self._name)
 
     def _add_sql_constraints(self):
-        """ Modify this model's database table constraints so they match the one
-        in _sql_constraints.
+        """ Modify this model's database table constraints and indexes
+        so they match the ones in _sql_objects.
+        """
+        for cons in self._sql_objects.values():
+            cons.sync_database_object(self)
 
+    @api.model
+    def __get_columns_from_sql_diagnostics(self, diagnostics, *, check_registry=False) -> list[str]:
+        """Given the diagnostics of an error, return the affected column names by the constraint.
+        Return an empty list if we cannot determine the columns.
         """
-        cr = self._cr
-        foreign_key_re = re.compile(r'\s*foreign\s+key\b.*', re.I)
-
-        for (key, definition, message) in self._sql_constraints:
-            conname = '%s_%s' % (self._table, key)
-            if len(conname) > 63:
-                hashed_conname = sql.make_identifier(conname)
-                current_definition = sql.constraint_definition(cr, self._table, hashed_conname)
-                if not current_definition:
-                    _logger.info("Constraint name %r has more than 63 characters, internal PG identifier is %r", conname, hashed_conname)
-                conname = hashed_conname
-            else:
-                current_definition = sql.constraint_definition(cr, self._table, conname)
-            if current_definition == definition:
-                continue
+        if column := diagnostics.column_name:
+            return [column]
+        if not check_registry:
+            return []
+        cr = self.env.cr
+        cr.execute(SQL("""
+            SELECT
+                conname,
+                t.relname,
+                ARRAY(
+                    SELECT attname FROM pg_attribute
+                    WHERE attrelid = conrelid
+                    AND attnum = ANY(conkey)
+                ) as "columns"
+            FROM pg_constraint
+            JOIN pg_class t ON t.oid = conrelid
+            WHERE conname = %s
+        """, diagnostics.constraint_name))
+        _constraint, constraint_table, fields = cr.fetchone() or (None, None, None)
+        if not fields or diagnostics.table_name != constraint_table:
+            return []
+        return fields
```

Comment:
small refactor (not exactly the same if the old SQL returns multi rows)
```suggestion
        cr.execute(SQL("""
            SELECT
                t.relname,
                ARRAY(
                    SELECT attname FROM pg_attribute
                    WHERE attrelid = conrelid
                    AND attnum = ANY(conkey)
                ) as "columns"
            FROM pg_constraint
            JOIN pg_class t
            ON t.oid = conrelid
            AND t.relname = %s
            AND conname = %s
        """, diagnostics.table_name, diagnostics.constraint_name))
        [fields] = cr.fetchone() or ([])
```